### PR TITLE
Add rigid-soft Divide-and-Truncate to SolverVBD

### PR DESCRIPTION
**Pull request by xiangdonglai**
## Description

Adds Divide-and-Truncate (DAT) penetration-avoidance for rigid–soft contact in `SolverVBD`, using the contact rows the collision pipeline already emits: particle–shape rows from `create_soft_contacts` and soft edge/face rows from `launch_soft_ef_contacts`. Each reported contact defines a division plane at its detection-time reference; rigid pose updates and particle displacements are truncated so neither side crosses it. Rigid motion is truncated along the curved trajectory of the stored surface point (sampling and bisection, with an optional interval-arithmetic pass), particles along their straight displacements, and both sides are truncated in both the rigid and particle phases. A per-body isotropic motion budget derived from the minimum rigid–soft query radius bounds motion between detections, DAT references are reset at collision-detection instants, and collision schedules are validated (the rigid slot may not be `NONE`; `AUTO` resolves to `PRE_POST_INIT`; both DAT families must share a schedule when active).

Soft self-contact DAT is moved onto the same machinery for consistency: division planes are built from certified primitive-pair separators (recomputed closest-point axis certified against the complete VT/EE pair, falling back to a sweep of candidate axes), truncation uses endpoint signs with a scale-adaptive float32 separation band, and the `[0.05, 0.95]` λ clamp additionally reserves a scale-adaptive float32 band on both sides when 5% of the gap is narrower. Wrong-side vertices may now recover but can no longer deepen; pairs with no strict separator (exactly touching or intersecting at the reference) are left unconstrained for that pass so contact forces can separate them. This is a behavior change to shipped self-contact DAT and is recorded in a `changed` fragment.

Also adds an opt-in C2 log-barrier rigid–soft normal contact law (`rigid_soft_contact_use_log_barrier`, default off). It is independent of DAT and can be evaluated on its own; it shares only the contact-force plumbing.

**Behavior changes to note (each has a Towncrier fragment):**
- Soft self-contact DAT now uses certified separators with endpoint-sign truncation and a scale-adaptive band; touching or self-intersecting pairs at a detection reference are left unconstrained for that pass (`changed`).
- `particle_conservative_bound_relaxation` is deprecated in favor of the common `dat_conservative_bound_relaxation`, which scales both self-contact and rigid-soft DAT truncation (`deprecated`).
- `step` raises when a DAT family's collision slot is `CollisionFrequencyType.NONE`: the `RIGID` slot with `rigid_soft_enable_dat`, and the `SOFT_SELF_CONTACT` slot with `particle_enable_self_contact`. The latter was previously accepted and silently ran without detection (its test `test_vbd_self_contact_none_starts_empty` is removed; the raise is covered in `test_solver_collision_frequency.py`) (`changed`).
- Post-initialization rigid detection (`PRE_POST_INIT`) previously read stale particle positions from `state_out`; it now uses the current `state_in` iterate (`fixed`).

New `SolverVBD` kwargs: `rigid_soft_enable_dat`, `rigid_soft_dat_use_interval_arithmetic`, `rigid_soft_contact_use_log_barrier`, and the common `dat_conservative_bound_relaxation` (shared by self-contact and rigid-soft DAT; `particle_conservative_bound_relaxation` is deprecated in its favor). New module `newton/_src/solvers/vbd/interval_arithmetic.py`. New example `vbd_dat_rigid_soft`.

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [x] For user-facing changes, a fragment has been added by following the
      [changelog fragment instructions](https://github.com/newton-physics/newton/blob/main/changelog/README.md)

## Test plan

```
uv run --extra dev -m newton.tests -k test_solver_vbd -k test_vbd_interval_arithmetic -k test_solver_collision_frequency -k test_solver_vbd_proxy_full_surface
uv run --extra dev -m newton.tests -k test_collision_cloth
uv run --extra dev -m newton.tests -k dat_rigid_soft
uv run -m newton.examples vbd_dat_rigid_soft --viewer null --test
```

- DAT unit and integration tests in `test_solver_vbd.py` (`TestVBDRigidDAT`): planar truncation endpoint-sign rules, rigid trajectory truncation (sampling/bisection and interval), division-plane placement extremes, config validation (owned pipeline, positive query radius, γ range, schedule NONE / mismatch), collision-frequency budget, analytic-sphere drop penetration bound with a DAT-off control, joint truncation in the rigid phase, soft self-contact exact-value truncation and fail-closed cases, and the log-barrier force law.
- `test_vbd_interval_arithmetic.py`: nextafter parity, interval-op containment, shortest-arc trig, signed-distance interval containment.
- Existing cloth self-contact and proxy full-surface suites pass unchanged.
- The `vbd_dat_rigid_soft` example asserts a `< 1e-4` m penetration bound over 300 frames in test mode.
- The example is registered in `test_examples.py` and in the README multiphysics gallery with a 320x320 screenshot; on CUDA devices it replays each frame as a captured graph like its multiphysics siblings.

## New feature / API change

```python
import newton

pipeline = newton.CollisionPipeline(model, broad_phase="nxn", soft_contact_gap=0.06)
solver = newton.solvers.SolverVBD(
    model,
    iterations=8,
    collision_pipeline=pipeline,      # DAT needs the solver-owned pipeline
    rigid_soft_enable_dat=True,       # truncate rigid poses and particles against per-contact planes
    dat_conservative_bound_relaxation=0.85,  # shared relaxation of truncation scalars and motion budgets
    rigid_soft_dat_use_interval_arithmetic=False,
    rigid_soft_contact_use_log_barrier=False,
)
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added opt-in Divide-and-Truncate support for rigid-soft contacts, including adaptive trajectory handling and optional interval verification.
  - Added an opt-in C2 log-barrier contact model for rigid-soft normal contacts.
  - Added a multiphysics example demonstrating penetration-free rigid-soft simulation.

- **Bug Fixes**
  - Improved truncation behavior for touching or unconstrained contact pairs.
  - Fixed post-initialization collision detection to use current particle positions.
  - Solver now reports invalid collision schedules instead of silently proceeding.

- **Documentation**
  - Documented new SolverVBD settings, collision scheduling, and shared DAT tuning controls.

### CODE REVIEW

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py [hunk 1]
```diff
@@ -2286,11 +2336,19 @@ def resolve_drive_limit_mode(
...
+                err_pos = q - lim_lower
         elif q > lim_upper:
-            mode = _DRIVE_LIMIT_MODE_LIMIT_UPPER
-            err_pos = q - lim_upper
+            if has_drive and drive_target < lim_upper:
+                mode = _DRIVE_LIMIT_MODE_DRIVE
+                err_pos = q - drive_target
+            else:
+                mode = _DRIVE_LIMIT_MODE_LIMIT_UPPER
+                err_pos = q - lim_upper
```
[the remark this answers is not in the document]
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/rigid_vbd_kernels.py]: Resolved by removing the change rather than documenting it: the drive/limit priority edit in `resolve_drive_limit_mode` was unrelated to DAT and had been carried in from the development branch. `5208ade3` restores main's limit-first behavior, so this PR no longer touches that function.

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 2]
```diff
@@ -319,13 +321,18 @@ def __init__(
...
         # Rigid body - joints
         rigid_joint_linear_ke: float = 1.0e5,  # Structural linear joint stiffness
         rigid_joint_angular_ke: float = 1.0e5,  # Structural angular joint stiffness
         rigid_joint_linear_k_start: float = 1.0e2,  # Legacy AVBD linear joint penalty ramp seed
         rigid_joint_angular_k_start: float = 1.0e1,  # Legacy AVBD angular joint penalty ramp seed
         rigid_joint_linear_kd: float = 0.0,  # Absolute damping for non-rod linear joint constraints
         rigid_joint_angular_kd: float = 0.0,  # Absolute damping for non-rod angular joint constraints
+        # Rigid body - penetration-free DAT truncation
+        rigid_soft_enable_dat: bool = False,  # Truncate rigid pose updates against per-contact division planes
+        rigid_soft_dat_relaxation: float = 0.85,  # Relaxation factor for rigid DAT truncation
```
Comment (xiangdonglai) [inline newton/_src/solvers/vbd/solver_vbd.py]: rigid_soft_dat_relaxation: reuse the soft self-contact relaxation
particle_conservative_bound_relaxation -> dat_conservative_bound_relaxation

Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Move to common parameters

Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Done in 96fcea15b: `dat_conservative_bound_relaxation` (common parameters, default 0.85) replaces `rigid_soft_dat_relaxation` and is shared with soft self-contact; `particle_conservative_bound_relaxation` is deprecated and overrides it with a `DeprecationWarning` through 1.7 (`deprecated` fragment).

File: newton/_src/solvers/vbd/particle_vbd_kernels.py [hunk 3]
```diff
@@ -2226,21 +2046,20 @@ def apply_planar_truncation_parallel_by_collision(
...
                 )
 
-                # For each, check the corresponding is_dummy entry in the vec4 is_dummy
-                if not is_dummy[0]:
-                    t = planar_truncation_t(e1_v1_pos, delta_e1_v1, n, d, parallel_eps, gamma)
+                # Without a strict separator at the reference (the edges intersect within
+                # float precision) DAT cannot tell which side is safe, so the pair is left
+                # unconstrained for this pass: freezing it would pin both edges at the
+                # reference indefinitely. Contact forces separate it and a later detection
+                # certifies a separator.
```
[the remark this answers is not in the document]
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/particle_vbd_kernels.py:2072]: Intentional; see the message of 96fcea15b. Freezing every vertex of an uncertifiable pair re-created the same detection reference on the next pass, so the pair was reported and frozen again indefinitely, including motion that separates it. That is a regression against main, which skips zero-normal VT pairs and lets EE pairs recover. Cloth has no inside or outside, so DAT cannot choose a side for a pair that is exactly touching or already intersecting at the reference; the truncation only prevents new crossings. Leaving such a pair unconstrained for the pass lets contact forces separate it, and the next detection certifies a separator.

Coverage: the zero-gap VT and intersecting EE subtests in `test_soft_self_dat_truncates_complete_primitive_pairs` assert that separating displacement passes untouched, and `test_soft_self_dat_touching_pair_separates` runs the full solver on a patch whose edge starts exactly inside another sheet and checks that it lifts off. All three fail on the fail-closed kernel.

Safety check: a 10 s cloth-twist run audited every substep with a float64 separating-axis test. 0 triangle crossings in 5/5 runs with this behavior, and 0/2 with the previous fail-closed kernel (the branch never fires in that scene), versus 4/5 runs crossing on main.

File: newton/examples/multiphysics/example_vbd_dat_rigid_soft.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    # rigid DAT penetration-free truncation (the feature under demonstration)
+    "enable_dat": True,
+    # contact detection gap: rigid DAT derives its per-detection motion budget
+    # (~0.5 * relaxation * gap) from this automatically; it must exceed the fastest
+    # per-detection-interval motion or DAT throttles the body
+    "soft_contact_gap": 0.06,
+    # body-body contact: compliant ALM with a stiff authored material. The default
+    # ShapeConfig stiffness (2.5e3 N/m) lets these 4-10 kg projectiles interpenetrate
+    # by several centimeters on impact; 3e6 N/m keeps overlap below 0.5 mm, while 1e7
+    # destabilizes the resting box at this substep and iteration budget.
```
[the remark this answers is not in the document]
Comment (xiangdonglai) [reply newton/examples/multiphysics/example_vbd_dat_rigid_soft.py]: Done in 4d3012920: the body-contact comment is down to the tuning rationale, and the two comments that restated the `top_cloth_*` names and the meaning of `camera_offset` are removed.

**Comment by xiangdonglai** [comment]
pre-commit.ci run

**Comment by xiangdonglai** [comment]
pre-commit.ci run

**Review by AnkaChan — COMMENTED** [review]
Reviewed in depth at b84c258d and re-anchored against the rebased head e5cf79df (the rebase over the #3995 soft-contact optimization did not change the DAT code my comments touch); I also ran the touched suites on an L40 (`test_solver_vbd` 167 OK, `test_collision_cloth` 37 OK, `TestVBDRigidDAT` 36 OK, `test_solver_collision_frequency` 34 OK, `test_vbd_interval_arithmetic` 6 OK, the 300-frame example), plus two checks the PR doesn't cover: CUDA graph capture of a DAT step works (capture + replays, zero penetration, matches eager), and the example scene costs 38 -> 52 ms/frame with DAT on (+37 %, includes the extra `PRE_POST_INIT` detection that `AUTO` adds).

No blocking findings. The design fits: it extends the existing self-contact DAT concept, the surface is four opt-in kwargs plus one deprecation, and the test coverage is thorough. My inline comments: one explicit acceptance of the touching-pair policy, one doc-contract fix for the interval module, and a few nits (error-message wording with suggestions, init ordering, dead parameter, README gallery entry).

Note for other readers: the self-contact truncation rework changes shipped behavior beyond what one fragment line suggests — certified separators, all four pair vertices constrained, endpoint signs instead of the parallel-eps ratio test, scale-adaptive band, fail-open for gap <= 0 pairs (`particle_edge_parallel_epsilon` no longer affects truncation, only detection/forces). The cloth suites pass unchanged.

File: newton/_src/solvers/vbd/particle_vbd_kernels.py [hunk 5]
```diff
@@ -2245,21 +2065,20 @@ def apply_planar_truncation_parallel_by_collision(
...
+                if valid_plane:
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/particle_vbd_kernels.py:2073]: Recording explicit acceptance of the fail-open policy here, since the CodeRabbit merge-risk note asks for it. I verified the livelock argument: freezing an uncertifiable pair zeroes the separating displacement each pass, so the next detection re-creates the identical touching reference and the pair stays pinned forever — `test_soft_self_dat_touching_pair_separates` reproduces it. With no certifiable separator there is also no trustworthy direction for a "no-deepen" rule, so leaving the pair to contact forces for the pass is the only option that cannot make things worse. The cost is correctly documented in the fragment (start simulations self-intersection-free). Accepted.

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py [hunk 6]
```diff
@@ -6849,3 +6905,768 @@ def update_rod_dahl_state(
...
+    s0 = wp.dot(n, rigid_point_trajectory(0.0, c0, dx, axis, angle, offset0) - d) - maximum_signed_distance
+    candidate_t = float(1.0)
+    crossed = bool(False)
+    if s0 > 0.0:
+        # Algorithm 1 assumes a valid starting half-space. A point already
+        # inside the forbidden band may move only monotonically away from it.
+        s_end = wp.dot(n, rigid_point_trajectory(1.0, c0, dx, axis, angle, offset0) - d) - maximum_signed_distance
+        # An endpoint-only test is insufficient for rotation: an arc can first
+        # worsen and then recover. Requiring the derivative to be nonpositive.
+        derivative_range = rigid_point_plane_signed_distance_derivative_interval(0.0, 1.0, n, dx, axis, angle, offset0)
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/rigid_vbd_kernels.py:7432]: `rigid_point_plane_signed_distance_derivative_interval` runs here for every wrong-side-start row regardless of `rigid_soft_dat_use_interval_arithmetic` — the flag only gates the Stage-2 prefix certification below. That contradicts the module docstring ("SolverVBD uses it only through an explicitly experimental Stage 2 option", interval_arithmetic.py:7-9) and the experimental framing in the kwarg docs and fragment.

The gate itself is right: for an arc, an endpoint test is strictly weaker than the "never deeper at any time" rule the straight-line particle path gets for free, and the returning-arc test pins that behavior in both modes. Please reconcile the contract — either document the derivative bound as a production dependency of the default path (accepting that the provisional 4-ulp trig padding backs default behavior), or repackage that one bound as a plain conservative closed-form helper in rigid_vbd_kernels.py so interval_arithmetic.py stays truly Stage-2-only. Docs-only is fine with me.

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 7]
```diff
@@ -772,6 +846,9 @@ def __init__(
         # Common parameters
         self.iterations = iterations
         self.friction_epsilon = friction_epsilon
+        self.rigid_soft_contact_use_log_barrier = bool(rigid_soft_contact_use_log_barrier)
+        # Needed by AUTO schedule resolution while rigid buffers are initialized.
+        self.rigid_soft_enable_dat = rigid_soft_enable_dat
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:851]: `_default_collision_frequency_type` reads `self.rigid_soft_enable_dat`, which is assigned here — after `super().__init__` (line 780) — and again in `_init_rigid_penetration_free` (line 2723). This only works because nothing resolves the RIGID slot in between and the SOFT_SELF_CONTACT resolution short-circuits on `slot == RIGID` first. Assign it before `super().__init__`, next to `particle_enable_self_contact` (line 763), which exists for exactly this reason, and drop one of the duplicates.

File: newton/examples/multiphysics/example_vbd_dat_rigid_soft.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 The Newton Developers
```
Comment (AnkaChan) [inline newton/examples/multiphysics/example_vbd_dat_rigid_soft.py:1]: The PR justifies omitting the README gallery entry as consistent with `newton/examples/vbd/*`, but the example now lives in `multiphysics/`, where several examples are in the gallery — including `example_rigid_soft_contact`, this example's nearest neighbor. Please add the gallery entry and 320x320 screenshot, or restate the omission for the new location.

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py [hunk 9]
```diff
@@ -6849,3 +6905,768 @@ def update_rod_dahl_state(
...
+        t = wp.clamp(wp.dot(point - segment_a, direction) / length_sq, 0.0, 1.0)
+        return segment_a + t * direction
+
+    reverse_direction = -direction
+    t = wp.clamp(wp.dot(point - segment_b, reverse_direction) / length_sq, 0.0, 1.0)
+    return segment_b + t * reverse_direction
+
+
+@wp.func
+def find_vertex_triangle_separator(
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/rigid_vbd_kernels.py:7065]: Both separator finders take a `normal_hint` as their last candidate axis, but every production call site passes `wp.vec3(0.0)` (e.g. particle_vbd_kernels.py:1946), which always fails the length check in `_certify_primitive_pair_separator`. Either feed the detection-time normal through or drop the parameter and the candidate slot.

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 10]
```diff
@@ -3383,7 +3792,7 @@ def _solve_particle_iteration(
                     ],
                     device=self.device,
                 )
-            self._penetration_free_truncation(state_in.particle_q)
+            self._penetration_free_truncation(state_in, contacts)
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:3795]: Measured on the new example at b84c258d, before the rebase (L40, 4 substeps x 8 iterations): 38 ms/frame DAT off -> 52 ms/frame DAT on (+37 %), including the second detection pass AUTO->PRE_POST_INIT adds. The joint truncation kernel runs `iterations * (particle_colors + 1) + 2` times per step. Opt-in, so no objection — but worth stating the expected cost in the example header or docs. Micro-nit: the self-contact kernel now issues `wp.atomic_min` even for `t = 1.0` results (particle_vbd_kernels.py:2074-2081), where the rigid kernel guards with `if t < 1.0`.

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 11]
```diff
@@ -3990,11 +4402,46 @@ def _resolve_self_contact_schedule(self):
...
+        return self._rigid_mode_this_step == Frequency.ITERATIONS and (iter_num + 1) % self._rigid_freq_this_step == 0
+
+    def _validate_dat_collision_schedules(self) -> None:
+        """Require active DAT families to have usable, mutually consistent schedules."""
+        Frequency = SolverBase.CollisionFrequencyType
+        if self.rigid_soft_enable_dat and self._rigid_mode_this_step == Frequency.NONE:
+            raise ValueError(
+                "rigid_soft_enable_dat requires an active rigid collision schedule; "
+                "collision_frequency_type NONE cannot maintain DAT's detection-centered "
+                "motion-bound and complete-pair invariants."
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:4422]: "detection-centered motion-bound and complete-pair invariants" is hard to parse and gives no fix. Suggested (keeps the substring the test matches):

```suggestion
                "rigid_soft_enable_dat requires an active rigid collision schedule: "
                "set the RIGID slot to PRE_INIT, PRE_POST_INIT, or ITERATIONS, not NONE."
```

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 12]
```diff
@@ -3990,11 +4402,46 @@ def _resolve_self_contact_schedule(self):
...
+            )
+        if self.particle_enable_self_contact and self._sc_mode_this_step == Frequency.NONE:
+            raise ValueError(
+                "particle_enable_self_contact requires an active soft self-collision schedule; "
+                "collision_frequency_type NONE cannot maintain DAT's detection-centered "
+                "motion-bound and complete-pair invariants."
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:4428]: Same here:

```suggestion
                "particle_enable_self_contact requires an active soft self-collision schedule: "
                "set the SOFT_SELF_CONTACT slot to PRE_INIT, PRE_POST_INIT, or ITERATIONS, not NONE."
```

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 13]
```diff
@@ -3990,11 +4402,46 @@ def _resolve_self_contact_schedule(self):
...
+            )
+        schedules_match = self._rigid_mode_this_step == self._sc_mode_this_step and (
+            self._rigid_mode_this_step != Frequency.ITERATIONS or self._rigid_freq_this_step == self._sc_freq_this_step
+        )
+        if self.rigid_soft_enable_dat and self.particle_enable_self_contact and not schedules_match:
+            raise ValueError(
+                "rigid-soft DAT and soft-self DAT share a particle trajectory reference and "
+                "therefore require equivalent rigid and soft self-collision schedules"
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:4436]: And here (keeps the "require equivalent" substring the test matches):

```suggestion
                "rigid-soft DAT and self-contact DAT share one motion reference and require "
                "equivalent RIGID and SOFT_SELF_CONTACT schedules (same type; same frequency for ITERATIONS)."
```

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 14]
```diff
@@ -2321,43 +2404,80 @@ def step(
...
-            self._run_rigid_collision(iterate, dt)
-            self._refresh_rigid_contact_state(contacts, refresh=True)
-            self._step_body_body_contact_frame(contacts, iterate.body_q, dt, 1.0, 1.0)
-            self._refresh_body_particle_contact_state(contacts, refresh=True)
+        # post init detection
+        rigid_collision_due = (
+            self.collision_pipeline is not None and self._rigid_mode_this_step == _Frequency.PRE_POST_INIT
+        )
+        soft_self_collision_due = (
+            self.particle_enable_self_contact and self._sc_mode_this_step == _Frequency.PRE_POST_INIT
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:2441]: `_sc_mode` is unclear, please use the full spelling

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 15]
```diff
@@ -2321,43 +2404,80 @@ def step(
...
             self._rigid_freq_this_step = self._collision_frequency[rigid_slot]
-            if rigid_mode in (_Frequency.PRE_INIT, _Frequency.PRE_POST_INIT, _Frequency.ITERATIONS):
-                # ITERATIONS' baseline includes the pre-init pass; in-loop
-                # re-detections run before iterations k, 2k, and so on.
-                self._run_rigid_collision(state_in, dt)
-                update_rigid = True
+
+        self._validate_dat_collision_schedules()
+
+        # pre-init detection
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:2412]: can we compress those logic into a separate function to make step's structure more clear.

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 16]
```diff
@@ -2321,43 +2404,80 @@ def step(
...
+        self._initialize_particles(state_in, state_out, contacts, dt)
 
-        if self._rigid_mode_this_step == _Frequency.PRE_POST_INIT:
-            iterate = self._rigid_iterate_view(state_in, state_out)
-            self._run_rigid_collision(iterate, dt)
-            self._refresh_rigid_contact_state(contacts, refresh=True)
-            self._step_body_body_contact_frame(contacts, iterate.body_q, dt, 1.0, 1.0)
-            self._refresh_body_particle_contact_state(contacts, refresh=True)
+        # post init detection
+        rigid_collision_due = (
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:2437]: same here as previous comment

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py [hunk 17]
```diff
@@ -1952,26 +1957,60 @@ def evaluate_rigid_contact_from_collision(
...
+    ``particle_vbd_kernels.evaluate_self_contact_force_norm``.  A zero contact
+    radius cannot define its positive-distance barrier interval, so it retains
+    the quadratic penalty law.
+    """
+    penetration_depth = collision_radius - distance
+    if not use_log_barrier or collision_radius <= 0.0:
+        return -k * penetration_depth, k
+
+    tau = collision_radius * 0.5
+    d_min = 1.0e-5
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/rigid_vbd_kernels.py:1979]: The C2 claim breaks when `collision_radius < 2e-5`: `d_min` is fixed at `1e-5` and can sit above `tau`, emptying the log-barrier region. At radius `1.5e-5`, crossing `d = d_min` jumps the normal force from `5e-6*k` (quadratic branch) to `5.625e-6*k` (extension branch) and the Hessian from `k` to `0.5625*k`. The particle self-contact law has the same latent behavior, so in practice this only bites sub-20-micrometer contact radii — but since this function is new and advertised as C2, please guard it (e.g. `d_min = wp.min(1.0e-5, 0.5 * tau)`) or document the radius floor. `test_rigid_soft_contact_log_barrier_matches_particle` only exercises radius 0.02.

File: newton/tests/test_solver_vbd.py [hunk 18]
```diff
@@ -5619,5 +5699,1251 @@ class TestVBDFullSurfaceContact(unittest.TestCase):
...
+    )
+    np.testing.assert_allclose(
+        state.particle_q.numpy(),
+        solver.pos_prev_collision_detection.numpy() + solver.particle_displacements.numpy(),
+        rtol=0.0,
+        atol=1.0e-7,
+    )
+
+
+class TestVBDRigidDAT(unittest.TestCase):
```
Comment (AnkaChan) [inline newton/tests/test_solver_vbd.py:6834]: Nothing in the PR exercises CUDA graph capture: no test creates a graph, and the example steps eagerly (most other multiphysics examples support graphs). I verified capture works today — the kernels use fixed launch dims with on-device count reads, and the DAT reference resets are device-side copies; capturing a full DAT step on this sphere-drop scene and replaying it 29x stayed penetration-free and matched the eager run exactly. But DAT is exactly the kind of feature (new per-step launches, reference resets, schedule-dependent host branching) where a capture regression would slip in silently. Please add a capture-and-replay regression test here, e.g. the sphere-drop scene with one step captured via `wp.ScopedCapture` and replayed.

**Review by AnkaChan — COMMENTED** [review]
Two more from continued verification: a C2 edge case in the new contact law, and a request for graph-capture regression coverage.

File: newton/_src/solvers/vbd/solver_vbd.py [hunk 19]
```diff
@@ -3604,6 +4014,9 @@ def _solve_rigid_body_iteration(
                 device=self.device,
             )
 
+        # Truncate the accumulated pose updates before the dual updates read them.
+        self._rigid_penetration_free_truncation(state_in, contacts)
```
Comment (AnkaChan) [inline newton/_src/solvers/vbd/solver_vbd.py:4018]: The rigid-phase truncation runs once per iteration, after the complete body color sweep — while the particle phase truncates per color. Between body colors, a fast body can therefore sit past its division plane transiently, and the remaining colors accumulate forces against that pose. It cannot NaN (below `d_min` the barrier switches to the C2 quadratic extension, and the default penalty is quadratic), but it feeds extremely stiff extension forces (Hessian `k2/d_min^2`) into the rest of the sweep, and the end-of-iteration clip then has to undo the excursion.

I suggest running the rigid truncation per body color as well, symmetric to the particle phase. The added cost is one `apply_rigid_soft_truncation` + `apply_body_truncation_ts` launch per body color (body color counts are small), and `apply_body_truncation_ts` only rewrites a pose when `t < 1`.

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py (hunk 17 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/rigid_vbd_kernels.py:1979]: Should I guard the same for soft self-contact? `d_min = wp.min(1.0e-5, 0.5 * tau)`

File: newton/_src/solvers/vbd/particle_vbd_kernels.py (hunk 5 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/particle_vbd_kernels.py:2074]: Thanks for recording the acceptance.

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py (hunk 6 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/rigid_vbd_kernels.py:7437]: Reconciled docs-only in b8eac3154: the `interval_arithmetic` module docstring, the `rigid_soft_dat_use_interval_arithmetic` kwarg doc and the tuning table now say the derivative (monotonicity) bound is part of the default rigid DAT path and only the Stage-2 prefix certification is experimental.

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 7 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Done in b8eac3154: `rigid_soft_enable_dat` is assigned before `super().__init__` next to `particle_enable_self_contact`; the post-init assignment and the `_init_rigid_penetration_free` parameter/re-assignment are removed.

File: newton/examples/multiphysics/example_vbd_dat_rigid_soft.py (hunk 8 above)
Comment (xiangdonglai) [reply newton/examples/multiphysics/example_vbd_dat_rigid_soft.py:1]: Added in b8eac3154: README multiphysics gallery cell + command cell and `docs/images/examples/example_vbd_dat_rigid_soft.jpg` (320x320); the PR description is updated accordingly.

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py (hunk 9 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/rigid_vbd_kernels.py:7066]: Kept, but documented in b8eac3154: `normal_hint` is reserved for the BVH-query rigid-soft path that follows this PR (it has a detection-time normal to offer); the SDF-row callers pass zero and the docstrings now say so.

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 10 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py:3742]: Both in b8eac3154. The example header now states the measured cost on an RTX 6000 Ada at 4x8: ~+3 % per frame with the frame replayed as a CUDA graph (23.8 -> 24.5 ms, within run-to-run noise) and ~+38 % without graph capture (35.9 -> 49.6 ms, host-bound small launches) - consistent with your L40 numbers. The self-contact truncation kernel now skips `atomic_min` for `t == 1`.

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 11 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Applied verbatim in b8eac3154.

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 12 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Applied verbatim in b8eac3154.

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 13 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Applied verbatim in b8eac3154 (the `require equivalent` substring the test matches is preserved).

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 14 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Renamed to `_self_contact_mode_this_step` / `_self_contact_freq_this_step` in b8eac3154, parallel to `_rigid_mode_this_step` / `_rigid_freq_this_step`.

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 15 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Done in b8eac3154: `step()` now calls `_pre_initialization_detection()`, and the post-init and in-iteration passes share `_mid_step_detection(..., preserve_history=...)` with `_collision_detection_due(phase_modes)` deciding which detectors run. Behavior is unchanged (`restore_history` was already `False` for the post-init pass).

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 16 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Covered by the same extraction in b8eac3154 (see the previous thread).

File: newton/_src/solvers/vbd/rigid_vbd_kernels.py (hunk 17 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/rigid_vbd_kernels.py]: Guarded in b8eac3154 with `d_min = wp.min(1.0e-5, 0.5 * tau)`, applied to both `_evaluate_rigid_soft_contact_force_norm` and `evaluate_self_contact_force_norm` so the two laws keep matching at every radius. `test_contact_barrier_c2_at_tiny_radius` checks force/Hessian continuity of both laws at radius 1.5e-5 across `d_min`, `tau` and the legacy `1e-5`; it fails without the guard. If you would rather leave the particle law untouched in this PR, say so and I will restrict it to the rigid law.

File: newton/tests/test_solver_vbd.py (hunk 18 above)
Comment (xiangdonglai) [reply newton/tests/test_solver_vbd.py:6954]: Added in b8eac3154: `test_rigid_dat_graph_capture_replays_match_eager` (CUDA) warms up two identical sphere-drop scenes, captures a two-substep step pair on one and replays it while the other steps eagerly, and asserts finite state, penetration-free cloth on both, progress of the replayed state, and agreement within 1e-2 m. On my RTX 6000 Ada the eager runs themselves differ run-to-run by ~1e-3 m (also with `DeterministicMode.RUN_TO_RUN`), so I could not assert bit-equality as you observed - if your setup was deterministic I would like to know what differs. The example now also captures a graph on CUDA like its multiphysics siblings.

File: newton/_src/solvers/vbd/solver_vbd.py (hunk 19 above)
Comment (xiangdonglai) [reply newton/_src/solvers/vbd/solver_vbd.py]: Implemented in b8eac3154: `_rigid_penetration_free_truncation` now runs after every body color inside `_solve_rigid_body_iteration` (the last color's call also feeds the dual updates). Cost note: the example has a single body color, so its frame time is unchanged (measured above); an FR3 has two colors and pays one extra truncation pass per iteration.

**Comment by AnkaChan** [comment]
Looks good to me. Approved.

**Review by AnkaChan — APPROVED** [review]
